Consolidate clipping parameters handling in a helper function

tik65536 opened the pull request:
Summary : 

- Added a helper function  `specify_clip_settings` to consolidate clipping parameters handling in each `grid_cell_xxx` function,  `grid_cellids_for_extent` and `cells_for_geo_points`  at one place.
- Inside the helper function, the `clipper_scale_factor` is added according to the resolution (above 16)
- The value of `clipper_scale_factor` comes from a static dictionary; those values come from the checking on `DGGRID` by @fmigneault  in the issue https://github.com/sahrk/DGGRID/issues/97#issuecomment-3642871117: 

 ```python
 rf_level_scalefactor = {17: "10000000", 18: "100000000", 
                                     19: "1000000000", 20: "10000000000"}
 ```

Related issue in DGGRID: 
https://github.com/sahrk/DGGRID/issues/97#issuecomment-3642871117

tik65536 committed:
added `z7string_to_z7int` function in `igeo7.py`

tik65536 committed:
added doc string to the `z7string_to_z7int` function.

tik65536 committed:
#44
- added `clipper_scale_factor` to Dggs class

tik65536 committed:
Revert "#44"

This reverts commit 693cf8bc9e13ead98353a432d22bcabd18f22e9d.

The fix is not done correctly according to https://github.com/allixender/dggrid4py/issues/44#issuecomment-3642796109

tik65536 committed:
#44
- followed the suggestion from https://github.com/allixender/dggrid4py/issues/44#issuecomment-3642796109
- added the helper function `specify_clip_setting` to handle clip[per] related metafile settings.
- But it failed the test case `test_cells_for_geo_points` with cell_ids_only=False.

tik65536 committed:
#44
- fixed the input_address_type is None in `cells_for_geo_points`
  when the original `output_address_type` is None and zone_id_only is False
- changed the value of rf_level_scalefactor dictionary to string instead of int
- added test cases for the clipped_scale_factor

tik65536 committed:
#44
-changed the helper function name from `specify_clip_setting` to `specify_clip_settings`

tik65536 committed:
#44
- change the type of clip_geom from AnyGeometry to "AnyGeometry"

fmigneault commented on dggrid4py/dggrid_runner.py:
```diff
@@ -855,6 +855,10 @@ def dgapi_grid_gen(
 
+        # clipper_scale_factor
+        if subset_conf.get("clipper_scale_factor"):
+            metafile.append("clipper_scale_factor " + subset_conf['clipper_scale_factor'])
```
Should it be added only when other `clip_...` subsetting configs were handled (above)?

It won't cause problems otherwise (parameter ignored), but it wouldn't uselessly pollute the metafile.

tik65536 replied:
Hi @fmigneault, thank you for the review. Yes, it should. I have updated the code for the comment.
To do so, I created an if block for handling the `clip_subset_type`, then applied the `clipper_scale_factor` at the end of the block. This avoids adding the scale factor at every `clip_subset_type` check. 

Merry X'mas and wish you a Happy New Year :smile:

fmigneault replied:
Thanks! Merry X'mas and Happy New Year to you too 🎅🎄🎁

fmigneault reviewed (commented):
Looks good.

tik65536 committed:
#44, #45
- grouped the handling of clip_subset_type under an if block
- moved the clipper_scale_factor under the if block
- removed the TYPE_CHECKING for AnyGeometry
- changed to use AnyGeometery dtype, instead of "AnyGeometry"

tik65536 committed:
#44, #45
bugfix syntax error when using the dictionary get.

tik65536 committed:
Merge branch 'allixender:master' into consolidate_clipping_parameters_handling_in_helper_function

tik65536 committed:
#45
- if "clipper_scale_factor" is specified in the conf_extra,
  it should included in the clip_metafile_settings.
